File: solutions/Java/NQueensIIBacktrackingOptimized.java
import java.util.*;

class Solution {
    public int totalNQueens(int n) {
        Set<Integer> cols = new HashSet<>();
        Set<Integer> diag1 = new HashSet<>();
        Set<Integer> diag2 = new HashSet<>();
        return backtrack(0, n, cols, diag1, diag2);
    }

    private int backtrack(int row, int n, Set<Integer> cols, Set<Integer> diag1, Set<Integer> diag2) {
        if (row == n) {
            return 1;
        }

        int count = 0;
        for (int col = 0; col < n; col++) {
            int d1 = row - col;
            int d2 = row + col;

            if (cols.contains(col) || diag1.contains(d1) || diag2.contains(d2)) {
                continue;
            }

            cols.add(col);
            diag1.add(d1);
            diag2.add(d2);

            count += backtrack(row + 1, n, cols, diag1, diag2);

            cols.remove(col);
            diag1.remove(d1);
            diag2.remove(d2);
        }

        return count;
    }
}

File: solutions/Python/n_queens_ii_backtracking_optimized.py
def totalNQueens(n: int) -> int:
    """Optimized backtracking: O(N!) time, O(N) space"""
    def backtrack(row, cols, diag1, diag2):
        if row == n:
            return 1
        count = 0
        for col in range(n):
            if col not in cols and (row - col) not in diag1 and (row + col) not in diag2:
                count += backtrack(row + 1, cols | {col}, diag1 | {row - col}, diag2 | {row + col})
        return count

    return backtrack(0, set(), set(), set())
